### ISSUE TITLE
New Issue with Application.Current.MainPage.DisplayAlert()?

### ISSUE THREAD

**Issue by LittleCornerDev** [jonathanpeppers/Xamarin.Forms.Mocks#63]
This documented usage of testing `Application.Current` used to work:

```
[SetUp]
public void SetUp()
{
    MockForms.Init();

    //This is your App.xaml and App.xaml.cs, which can have resources, etc.
    Application.Current = new App();
}

[TearDown]
public void TearDown()
{
    Application.Current = null;
}
```

With latest version of Xamarin.Forms 4.8.0.1560 (only caught it now, not sure if it happened at earlier version), it seems to no longer be sufficient and causes the error:
```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
``` 
from a line that calls `Application.Current.MainPage.DisplayAlert()`.

Do we need to mock out `Application.Current` in a new way?  The `Application.Current.MainPage.DisplayAlert()` is called within `new App()` if errors are encountered upon app initialization, so unclear if/how `Application.Current = new App();` should be changed. Should we just stub out `Application.Current` to our own mock?

**Comment by jonathanpeppers** [comment]
> System.NullReferenceException: Object reference not set to an instance of an object.

What is the stack trace for this? You should see a list of method names, and maybe line numbers?

Are your tests running in parallel like in https://github.com/jonathanpeppers/Xamarin.Forms.Mocks/issues/60#issuecomment-707264326? That won't work because of the way Xamarin.Forms APIs use static variables.

**Comment by LittleCornerDev** [comment]
I do not know if the tests have been running in parallel this whole time.  I just know `dotnet` wasn't crashing before.

As previously mentioned the crash happens on the line in code where `DisplayAlert()` is called, but here is the stack trace:

```
Microsoft (R) Test Execution Command Line Tool Version 16.7.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at PROJECT.Utilities.UserUtil.LogIn() in PROJECT/Utilities/UserUtil.cs:line 198
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

I will investigate further to see if I can isolate the error further and provide more info.

**Comment by jonathanpeppers** [comment]
Your example shows that either `Current` or `MainPage` is null.

What test framework are you using, and how are you running these tests? `dotnet test`?

**Comment by LittleCornerDev** [comment]
Yes, using `dotnet test` (via a Custom Command) to run the nUnit test project automatically when doing a Debug/Release build on the Xamarin.Forms project.

Setting `Application.Current = new App();` as per `Xamarin.Forms.Mocks` documentation.

**Comment by jonathanpeppers** [comment]
Did you try something like this to see if it's caused by tests running in parallel?

https://github.com/nunit/nunit3-vs-adapter/issues/657#issuecomment-564313415

**Comment by LittleCornerDev** [comment]
As per suggestion, just ran `dotnet test -p:BuildInParallel=false <UNIT TEST PROJECT NAME>.csproj`, but it unfortunately had the same error.

The same error also happens when running the unit test project from Visual Studio's run/build button.  Commenting out that DisplayAlert() line runs all the tests successfully again.

**Comment by LittleCornerDev** [comment]
Ran every test one by one and found the unit test that's specifically causing the issue.  Commenting it out also makes all tests run fine.  It was recently moved from another file, so I think this is an issue with how the unit test was refactored and not `Xamarin.Forms.Mocks`.  Thanks for all your assistance!  Closing this inquiry.
